ion/symbols_table.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package ion
+
+type symbolTableKind uint8
+
+const (
+	symbolTableKindSystem symbolTableKind = iota
+	symbolTableKindShared
+	symbolTableKindLocal
+)
+
+// SymbolTable is the core lookup structure for Tokens from their symbolic ID
+type SymbolTable struct {
+	kind    symbolTableKind
+	name    string
+	version int32
+	imports []SymbolTable
+	textMap map[string]SymbolToken
+	tokens  []SymbolToken
+	maxSID  int64
+}
+
+func newSymbolTableRaw(kind symbolTableKind, name string, version int32) *SymbolTable {
+	table := SymbolTable{
+		kind:    kind,
+		name:    name,
+		version: version,
+		textMap: make(map[string]SymbolToken),
+	}
+	return &table
+}
+
+// newSymbolTable constructs a new empty symbol table of the given type
+// Imports are the tables to load into this one.
+func newSymbolTable(kind symbolTableKind, name string, version int32, imports ...SymbolTable) *SymbolTable {
+	table := newSymbolTableRaw(kind, name, version)
+	// copy in the import "references"
+	table.imports = append(imports[:0:0], imports...)
+	if kind == symbolTableKindLocal {
+		// for local symbol table, import system symbol table
+		imports = append([]SymbolTable{systemSymbolTable}, imports...)
+	}
+	for _, importTable := range imports {
+		// TODO Consider if we should not inline the imports or make this delegate to the imports or configurable
+		table.tokens = append(table.tokens, importTable.tokens...)
+		for text, newToken := range importTable.textMap {
+			_, exists := table.textMap[text]
+			if !exists {
+				table.textMap[text] = newToken
+			}
+		}
+		table.maxSID += importTable.maxSID
+	}
+	return table
+}
+
+// newLocalSymbolTable creates an instance with symbolTableKindLocal.
+// The name is empty and the version is zero, these fields are inapplicable to local symbol tables.
+func newLocalSymbolTable(imports ...SymbolTable) *SymbolTable {
+	return newSymbolTable(symbolTableKindLocal, "", 0, imports...)
+}
+
+// newSharedSymbolTable creates an instance with symbolTableKindShared.
+// returns `nil` if the version is not positive.
+func newSharedSymbolTable(name string, version int32, imports ...SymbolTable) *SymbolTable {
+	if version <= 0 {
+		return nil
+	}
+	return newSymbolTable(symbolTableKindShared, name, version, imports...)
+}
+
+// BySID returns the underlying SymbolToken by local ID.
+func (t SymbolTable) BySID(sid int64) (SymbolToken, bool) {
```

Comment:
If it is required to use a "constructor" to build an instance, and it returns a pointer to the struct, then all of the receivers should be pointer receivers: `(t *SymbolTable)` for consistency.

Reply:
Done.